server/src/main/java/org/apache/druid/discovery/BaseNodeRoleWatcher.java

```diff
@@ -209,13 +217,52 @@ private void removeNode(DiscoveryDruidNode druidNode)
     }
   }
 
+  public void cacheInitializedTimedOut()
```

Comment:
This patch has a similar issue to the one described here: https://github.com/apache/druid/pull/15726#pullrequestreview-1875684108.

It is relying on `getAllNodes()` to be called in order to detect that initialization has timed out. But `getAllNodes()` is not necessarily ever called. There needs to be some backstop that causes initialization to time out even if `getAllNodes()` is never called. Please add a test for this case as well.

You will probably need a `ScheduledExecutorService` to make this happen. It could be owned by the `CuratorDruidNodeDiscoveryProvider` and shared across all watchers. I think it would work to change the `ExecutorService listenerExecutor` to a `ScheduledExecutorService` and use that.

Reply:
Sorry, I've made the changes